archive/import.go

```diff
@@ -161,48 +161,101 @@ func (d *importDriver) close() error {
 	return idxWriter.Close()
 }
 
-func (d *importDriver) Write(cid int, batch zbuf.Batch) error {
-	if cid != 0 {
-		panic("importDriver write to non-zero channel")
-	}
+func (d *batchWriter) Write(batch zbuf.Batch) error {
 	for i := 0; i < batch.Length(); i++ {
 		if err := d.writeOne(batch.Index(i)); err != nil {
 			return err
 		}
 	}
+	err := d.close()
 	batch.Unref()
-	return nil
+	return err
 }
 
-func (d *importDriver) ChannelEnd(cid int) error {
-	if cid != 0 {
-		panic("importDriver ChannelEnd to non-zero channel")
+func importCompareFn(ark *Archive) expr.CompareFn {
+	return func(a, b *zng.Record) (cmp int) {
+		d := a.Ts() - b.Ts()
+		if d < 0 {
+			cmp = -1
+		}
+		if d > 0 {
+			cmp = 1
+		}
+		return cmp * ark.DataSortDirection.Int()
 	}
-	return d.close()
 }
 
-func (d *importDriver) Warn(warning string) error          { return nil }
-func (d *importDriver) Stats(stats api.ScannerStats) error { return nil }
-
-func importProc(ark *Archive) string {
-	if ark.DataSortDirection == zbuf.DirTimeForward {
-		return "sort ts"
+func Import(ctx context.Context, ark *Archive, zctx *resolver.Context, r zbuf.Reader, maxBufSize int64) error {
+	if maxBufSize == 0 {
+		maxBufSize = DefaultImportBufSize
 	}
-	return "sort -r ts"
-}
-
-func Import(ctx context.Context, ark *Archive, zctx *resolver.Context, r zbuf.Reader) error {
-	proc, err := zql.ParseProc(importProc(ark))
-	if err != nil {
-		return err
+	if maxBufSize < ark.LogSizeThreshold {
+		maxBufSize = ark.LogSizeThreshold
 	}
-	id := &importDriver{
+	bw := &batchWriter{
 		ark:            ark,
 		ctx:            ctx,
 		needIndexWrite: true,
 	}
-	if err := driver.Run(ctx, id, proc, zctx, r, driver.Config{}); err != nil {
-		return fmt.Errorf("archive.Import: run failed: %w", err)
+	bsort := newBoundedSort(r, importCompareFn(ark), maxBufSize)
+	for {
+		batch, err := bsort.Batch(ctx)
+		if err != nil {
+			return err
+		}
+		if batch == nil {
+			break
+		}
+		if err := bw.Write(batch); err != nil {
+			return err
+		}
 	}
 	return nil
 }
+
+type boundedSort struct {
+	maxsize   int64
+	reader    zbuf.Reader
+	compareFn expr.CompareFn
+}
+
+func newBoundedSort(reader zbuf.Reader, compareFn expr.CompareFn, maxsize int64) *boundedSort {
+	return &boundedSort{
+		maxsize:   maxsize,
+		reader:    reader,
+		compareFn: compareFn,
+	}
+}
+
+// Batch reads the next Batch into memory from the provided reader and returns the
+// next Batch when either the reader reaches EOS, the size of the batch reaches
+// maxsize, or the context errors out.
+func (b *boundedSort) Batch(ctx context.Context) (zbuf.Batch, error) {
+	var size int64
+	// XXX At some point we should use a memory pool to reduce allocations.
+	array := zbuf.Array{}
+	for {
```

Comment:
If you move this loop into the existing loop of `Import`, you could:
- use an `var arr []*zng.Record` variable directly instead of a zbuf.Batch, and reset it (`arr = arr[:0]`) between writes
- you could loop over the array and call `writeOne` directly, or just adjust the writer to take an `[]*zng.Record` parameter.